File: androidsupport/src/main/java/com/jarklee/androidsupport/utils/permission/PermissionRequest.java
/*
 * ******************************************************************************
 *  Copyright Ⓒ 2017. TrinhQuan. All right reserved
 *  Author: TrinhQuan. Created on 2017/1/13
 *  Contact: trinhquan.171093@gmail.com
 * ******************************************************************************
 */

package com.jarklee.androidsupport.utils.permission;

final class PermissionRequest {

    private final Runnable mActionOnGranted;
    private final Runnable mActionOnCanceled;
    private final boolean mIsRequireAll;

    private PermissionRequest(final Runnable actionOnGranted,
                              final Runnable actionOnCanceled,
                              boolean requireAll) {
        this.mActionOnCanceled = actionOnCanceled;
        this.mActionOnGranted = actionOnGranted;
        this.mIsRequireAll = requireAll;
    }

    void executeAction() {
        if (mActionOnGranted != null) {
            mActionOnGranted.run();
        }
    }

    void executeCancel() {
        if (mActionOnCanceled != null) {
            mActionOnCanceled.run();
        }
    }

    boolean requireAll() {
        return mIsRequireAll;
    }

    static PermissionRequest newGrantOnly(final Runnable actionOnGranted,
                                          final boolean requireAll) {
        return new PermissionRequest(actionOnGranted, null, requireAll);
    }

    static PermissionRequest newCancelOnly(final Runnable actionOnCanceled,
                                           final boolean requireAll) {
        return new PermissionRequest(null, actionOnCanceled, requireAll);
    }

    static PermissionRequest newRequest(final Runnable actionOnGranted,
                                        final Runnable actionOnCanceled,
                                        final boolean requireAll) {
        return new PermissionRequest(actionOnGranted, actionOnCanceled, requireAll);
    }

}

File: androidsupport-kotlin/src/main/java/com/jarklee/androidsupport/utils/permission/PermissionRequest.kt
/*
 * ******************************************************************************
 *  Copyright Ⓒ 2017. TrinhQuan. All right reserved
 *  Author: TrinhQuan. Created on 2017/1/14
 *  Contact: trinhquan.171093@gmail.com
 * ******************************************************************************
 */

package com.jarklee.androidsupport.utils.permission

internal class PermissionRequest private constructor(private val mActionOnGranted: PermissionActionCallback?,
                                                     private val mActionOnCanceled: PermissionActionCallback?,
                                                     private val mIsRequireAll: Boolean) {

    fun executeAction() {
        mActionOnGranted?.invoke()
    }

    fun executeCancel() {
        mActionOnCanceled?.invoke()
    }

    fun requireAll(): Boolean {
        return mIsRequireAll
    }

    companion object {

        fun newGrantOnly(actionOnGranted: PermissionActionCallback?,
                         requireAll: Boolean): PermissionRequest {
            return PermissionRequest(actionOnGranted, null, requireAll)
        }

        fun newCancelOnly(actionOnCanceled: PermissionActionCallback?,
                          requireAll: Boolean): PermissionRequest {
            return PermissionRequest(null, actionOnCanceled, requireAll)
        }

        fun newRequest(actionOnGranted: PermissionActionCallback?,
                       actionOnCanceled: PermissionActionCallback?,
                       requireAll: Boolean): PermissionRequest {
            return PermissionRequest(actionOnGranted, actionOnCanceled, requireAll)
        }
    }
}